buildsys: extend `external-files` to vendor go modules

jpmcb opened the pull request:
### Issue number:

Closes #2052
Closes #2053

### Description of changes:

```
buildsys: extend external-files to vendor go modules

Using the `bundle-*` keys on
`packag.metadata.build-package.external-files`, buildsys can vendor Go
dependencies into a new archive (leaving the prestine upstream archive
intact). Both archives may then be used to build the package in an
isolated environment.

toosl/docker-go now also retrieves the `GOSUMDB` environment variable
from the host.

Co-authored-by: John McBride <jpmmcb@amazon.com>
Co-authored-by: Samuel Karp <skarp@amazon.com>
Signed-off-by: John McBride <jpmmcb@amazon.com>

---

hotdog: use new buildsys go module support

Co-authored-by: John McBride <jpmmcb@amazon.com>
Co-authored-by: Samuel Karp <skarp@amazon.com>
Signed-off-by: John McBride <jpmmcb@amazon.com>

---

oci-add-hooks: use new buildsys go module support

Co-authored-by: John McBride <jpmmcb@amazon.com>
Co-authored-by: Samuel Karp <skarp@amazon.com>
Signed-off-by: John McBride <jpmmcb@amazon.com>
```

This patch carries many of the changes from #2053 but adds recommendations from subsequent code reviews.

The `external-files` interface has been extended to now take several optional `bundle-* modifiers that enable it to consume an upstream archive, unpack it, vendor it's go module dependencies, and create a new archive with those vendored go dependencies.

Further, this enables a possible future paradigm of vendoring _many_ modules from a single `external-files`. With this PR, only "go" is supported in `bundled-modules`.

**Testing done:**

Builds 👍🏼 
```
cargo make -e PACKAGE=oci-add-hooks build-package
cargo make -e PACKAGE=hotdog build-package
```

and launching an ecs instance looks 👍🏼 
```
cargo make -e BUILDSYS_VARIANT=aws-ecs-1
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

jpmcb commented:
~Whoops! A few files need `rustfmt` - please hold!~

Force pushed to address failing GitHub actions needing `rustfmt`-ing

jpmcb commented:
Force pushed to address `cargo make check-lints` and `cargo make check-fmt`

jpmcb commented:
Force pushed to address not having to unwrap here:

```diff
- dg_args
-         .go_mod_cache 
-         .to_str()
-         .context(error::InputFileSnafu)
-         .unwrap()
+ dg_args
+        .go_mod_cache
+        .to_str()
+        .context(error::InputFileSnafu)?
```

arnaldo2792 commented on tools/buildsys/src/gomod.rs:
```diff
@@ -0,0 +1,176 @@
+/// and ensure that we track it for changes, since it will directly affect the
+/// output.
+fn getenv(var: &str) -> Result<String> {
+    println!("cargo:rerun-if-env-changed={}", var);
```
I think we are not supposed to do this anymore and instead use `bottlerocket-variant`, and probably this function? https://github.com/bottlerocket-os/bottlerocket/blob/037d0e2798a7e8d4a12d0fb35d3c266db3d74f25/sources/bottlerocket-variant/src/lib.rs#L178

I'll let others comment on this since I'm not an expert about when the library linked should be used.

jpmcb replied:
Good question: This was a chunk of code from Sam's initial PR and I assumed would force a rerun if any of the go mod control env variables changed. If there's a better way to do this, let me know!

arnaldo2792 commented on tools/buildsys/src/gomod/error.rs:
```diff
@@ -0,0 +1,42 @@
+    #[snafu(display("Failed to execute command: 'docker {:?}'", args))]
+    DockerExecution { args: String },
+
+    #[snafu(display("input url is required"))]
```
Nit: `s/input/Input`, same for error below

arnaldo2792 commented on tools/buildsys/src/gomod/error.rs:
```diff
@@ -0,0 +1,42 @@
+    #[snafu(display("output-dir is required"))]
+    OutputDir,
+
+    #[snafu(display("output_dir must not exist or be a directory"))]
+    OutputDirBad,
```
This error isn't used, you could remove it.

jpmcb replied:
Good catch - thanks!

arnaldo2792 commented on tools/buildsys/src/manifest.rs:
```diff
@@ -33,6 +33,35 @@ url = "https://bar"
+The `bundle-*` keys on `external-files` are a group of optional modifiers
+and are used to untar an upstream external file achive,
+vendor any dependent code, and bundle it back up to be consumed.
+All three options are required when bundling an archieves dependences.
```
Nit: missing `'` in `archives`?

stmcginnis replied:
s/archieves dependences/archive's dependencies/

stmcginnis commented on tools/buildsys/src/manifest.rs:
````diff
@@ -33,6 +33,35 @@ url = "https://bar"
 ```
 
+The `bundle-*` keys on `external-files` are a group of optional modifiers
+and are used to untar an upstream external file achive,
````
s/achive/archive/

stmcginnis reviewed (commented):
Since it looks like you will need to make other updates, added a couple small nits to the list.

stmcginnis commented on tools/buildsys/src/gomod.rs:
```diff
@@ -0,0 +1,176 @@
+                export GOPROXY={goproxy} &&
+                export GOSUMDB={gosumdb} &&
+                export GOPRIVATE={goprivate} &&
+                go list -mod=readonly ./... >/dev/null && go mod vendor &&
```
This `go mod vendor` seems to be causing problems for me. For projects that just use `go.mod`, this causes a `vendor` directory to be created and everything to be downloaded (again) into that new directory. That seems to be the cause of these errors:

```
  #19 0.532 go: inconsistent vendoring in /home/builder/rpmbuild/BUILD/cloud-provider-aws-1.22.3/GOPATH/src/github.com/kubernetes/cloud-provider-aws:
  #19 0.532     github.com/aws/aws-sdk-go@v1.43.32: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
  #19 0.532     github.com/golang/mock@v1.4.4: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
  #19 0.532     github.com/google/go-cmp@v0.5.5: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
  #19 0.532     github.com/stretchr/testify@v1.7.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
...
```

Can we add a check somewhere that first checks if there is a `vendor` directory, and based on that decide whether to run `go mod vendor` or not?

stmcginnis replied:
Maybe this isn't the actual source of the failure. I tried removing this locally, but I still get the error. So my error may be coming from elsewhere, or it may be left over cached files in my environment. If it's something left behind, I haven't been able to figure out yet where that would be or how to clean it up though.

But somewhere a `go mod vendor` is happening and seems to be the source of my problems.

jpmcb replied:
> Can we add a check somewhere that first checks if there is a vendor directory, and based on that decide whether to run go mod vendor or not?

If there is already an existing `vendor` directory, we shouldn't be calling out to `go mod` again. This would indicate that the upstream project has declared their dependencies, has them part of their source control, and the project is ready to be build via `go build`. I _believe_ this is already the current paradigm as evident by packages like `runc`: 

https://github.com/bottlerocket-os/bottlerocket/blob/211e3db67a7c0ac758c6eb135213285513c92f82/packages/runc/runc.spec#L34-L38

and their vendoring inside the project

https://github.com/opencontainers/runc/tree/main/vendor

stmcginnis replied:
> If there is already an existing vendor directory, we shouldn't be calling out to go mod again

No, my point was if there ISN'T an existing `vendor` directory, then don't try to run `go mod vendor` as that's not the type of dependency management that project uses. But based on our offline conversation, sounds like this may still be wanted to be a way to "cache" those dependencies locally to allow things to be built in a network isolated container.

So maybe this is just a red herring and there's some other issue with the package I'm working on. Still trying to sort that out.

jpmcb replied:
> may still be wanted to be a way to "cache" those dependencies locally to allow things to be built in a network isolated container.

Right, this will enable packages being build in our network isolated `buildsys` container to not have to gather the dependencies manually (like hotdog _was_ doing) but rather, build a "bundled" tarball with the vendored dependencies to then be consumed in the isolated environment

jpmcb commented:
Force pushed to address nits / spelling mistakes ✏️

bcressey commented on packages/hotdog/hotdog.spec:
```diff
@@ -10,67 +10,32 @@
-Source3: https://github.com/opencontainers/selinux/archive/refs/tags/v%{goselinux}.tar.gz#/go-selinux-v%{goselinux}.tar.gz
-Source4: https://git.kernel.org/pub/scm/libs/libcap/libcap.git/snapshot/libcap-cap/v%{libcap}.tar.gz#/libcap-v%{libcap}.tar.gz
-
+Source0: %{gorepo}-%{shortrev}-bundled.tar.gz
```
I'd like to have two source files in this case:
* Source0 - pristine upstream archive
* Source1 - just the vendored modules

jpmcb replied:
> I'd like to have two source files in this case:

Hmmm, what is the benefit of having a separate untouched upstream archive alongside it's vendored modules?

In this PRs current state, the archive is downloaded, unpacked, modified via `go mod vendor` (which populates it's dependencies into a `vendor` directory) and then tar'd back up into a file named by `bundle-output-path`. This gives a simple, **single file** interface to then use in the build spec. It also emulates what a Go project _would_ look like if it had vendored it's modules into the source control.

But, I suppose with separate archives for the upstream source and bundled modules, we could have greater control over what makes it into end result. I.e., if we needed to patch an upstream's dependency at build time, we'd want to do that out of band from the actual upstream source code which would be painful if it's all already packaged into one archive.  

I guess this is really about which piece of the package building process should we combine the vendored modules and the upstream source: in the lookaside cahce's archive? After `buildsys` has downloaded the archive into a single tar file that gets passed to the spec? Within the spec itself during build time?

Seems like for the most control, we'd want the last option, at buildtime?

bcressey replied:
> Hmmm, what is the benefit of having a separate untouched upstream archive alongside it's vendored modules?

It feels surprising to have `Cargo.toml` list one file with its hash, and have the spec file switch over to using a different file with an unknown hash without any warning. If you don't know about the magic that `buildsys` is doing, you might think the hash isn't actually being used, which would suggest a security problem.

By having the bundle archive just cover the delta - e.g. the vendor directory - you don't run the risk of something happening to one of the source files that isn't expected to change, which would then be difficult to track down.

> Seems like for the most control, we'd want the last option, at buildtime?

This seems right to me, even if it means a second `%setup` invocation to extract the second archive.

jpmcb replied:
> I'd like to have two source files in this case

👍🏼 Two archives are now used as `Source0` (the upstream pristine archive) and `Source1` which has the vendored dependencies

bcressey commented on tools/buildsys/src/manifest.rs:
```diff
@@ -33,6 +33,35 @@ url = "https://bar"
+"go" may be entered to vendor the dependencies.
+Currently, only "go" is supported.
+
+`bundle-path` is the filepath _within_ the archive that contains the module.
+So, for example, given a go project that has the necessary go.mod and go.sum
+files within the archive located at the filepath a/b/c,
+this `bundle-path` value should be `a/b/c`.
```
Can we default this to the first sub-directory in the archive, assuming the root level is otherwise empty?

jpmcb replied:
👍🏼 Now defaults to the first top level directory in the archive if `bundle-path` is not provided

bcressey commented on tools/buildsys/src/manifest.rs:
```diff
@@ -33,6 +33,35 @@ url = "https://bar"
+files within the archive located at the filepath a/b/c,
+this `bundle-path` value should be `a/b/c`.
+
+`bundle-output-path` is the path of the output tar archive.
+This name may be referenced within a spec, when creating a package, etc.
+For example, `my-package-bundled.tar.gz` may be entered as an output path
+indicating the package is "bundled" and has been archived.
```
A default value using a convention where `-bundled` is appended to the last part of the external filename before `.tar.*` would be nice.

jpmcb replied:
👍🏼 I punted on this _abit_ since I was having trouble with Rust string manipulation. It now defaults to `bundled-{name-of-tar}`. So, for example, in the `hotdog` package, the default would be `bundled-hotdog-b85b755.tar.gz` if `bundled-output-path` is not provided.

bcressey commented on tools/buildsys/src/manifest.rs:
```diff
@@ -352,6 +381,9 @@ pub(crate) struct ExternalFile {
     pub(crate) path: Option<PathBuf>,
     pub(crate) sha512: String,
     pub(crate) url: String,
+    pub(crate) bundle_modules: Option<Vec<String>>,
```
We can enforce stronger typing here using an enum, like `SupportedArch` and `GrubFeature` do.

(Then use that enum with a match statement in `main.rs` to ensure that all known values are handled.)

jpmcb replied:
👍🏼  Added enum for this struct field

bcressey commented on tools/buildsys/src/gomod.rs:
```diff
@@ -0,0 +1,176 @@
+        // Our SDK and toolchain are picked by the external `cargo make` invocation.
+        let sdk = getenv("BUILDSYS_SDK_IMAGE")?;
+
+        // Several Go variables control proxying
+        let mut goproxy = go_env("GOPROXY").unwrap_or_else(|| "".to_string());
+        if goproxy.ends_with('\n') {
+            goproxy.pop();
+        }
+        let mut gosumdb = go_env("GOSUMDB").unwrap_or_else(|| "".to_string());
+        if gosumdb.ends_with('\n') {
+            gosumdb.pop();
+        }
+        let mut goprivate = go_env("GOPRIVATE").unwrap_or_else(|| "".to_string());
+        if goprivate.ends_with('\n') {
+            goprivate.pop();
+        }
```
nit: maybe factor this into a helper function that takes a list of possible environment variables

jpmcb replied:
👍🏼 This code block was redundant since `docker-go` gets these variables from the environment so this block was removed.

bcressey commented on tools/buildsys/src/gomod.rs:
```diff
@@ -0,0 +1,176 @@
+    Ok(output)
+}
+
+/// Run `go env` with the specified argument.
+fn go_env(var: &str) -> Option<String> {
+    let args = vec!["env", var];
+    let output = match cmd("go", args)
+        .stderr_to_stdout()
+        .stdout_capture()
+        .unchecked()
+        .run()
+    {
+        Ok(v) => v,
+        Err(_) => return None,
+    };
+    let stdout = String::from_utf8_lossy(&output.stdout);
+    println!("{}", &stdout);
+    output.status.success().then(|| stdout.to_string())
+}
```
We don't want to assume that a `go` install exists on the host or works as expected.

jpmcb replied:
Gotcha - what would good indicators that buildsys shouldn't proceed with vendoring the go module on the host?

My first thought is to check:
- is the `go` tool in the path?
- is the output of `go version` greater than 1.16?

If either of the answer to those is "no", we fail and exit early.

bcressey replied:
Hmm, is this redundant with the checks in `docker-go`? That script already has the logic to use `go` if we have it, otherwise read the environment variable directly.

jpmcb replied:
👍🏼 Removed this block since `docker-go` is doing the same thing and gets the necessary go env vars

bcressey commented on tools/buildsys/src/gomod.rs:
```diff
@@ -0,0 +1,176 @@
+    output.status.success().then(|| stdout.to_string())
+}
+
+/// Retrieve a BUILDSYS_* variable that we expect to be set in the environment,
+/// and ensure that we track it for changes, since it will directly affect the
+/// output.
+fn getenv(var: &str) -> Result<String> {
+    println!("cargo:rerun-if-env-changed={}", var);
+    env::var(var).context(error::EnvironmentSnafu { var })
+}
```
`BUILDSYS_SDK_IMAGE` will be tracked by the usage of `getenv` in `builder.rs`, so we don't need it here.

jpmcb replied:
👍🏼 Removed this function and get `BUILDSYS_SDK_IMAGE` directly. Omitted the `rerun-if-env-changed` call

bcressey commented on tools/buildsys/src/gomod.rs:
```diff
@@ -0,0 +1,176 @@
+            .bundle_output_path
+            .as_ref()
+            .context(error::OutputDirSnafu)?;
+        let output_path = package_dir.join(output_path_arg);
```
Somewhere, maybe right here, we need to emit `rerun-if-changed` for the output bundle archive.

jpmcb replied:
👍🏼 Done!

bcressey commented on tools/buildsys/src/gomod.rs:
```diff
@@ -0,0 +1,176 @@
+            .context(error::OutputDirSnafu)?;
+        let output_path = package_dir.join(output_path_arg);
+
+        // Return early if the output path exists and is a file, assuming it's already been built
+        if output_path.exists() && output_path.is_file() {
+            return Ok(());
+        }
```
The discussion on single vs separate archives made me think of a potential failure mode here, which is that if the input external file changes in some way but keeps the same filename, we need to detect that and recreate the dependency bundle, rather than reusing the one we have.

If this optimization doesn't save very much time, it might be simpler to just always recreating the bundle file. I'm assuming that the `go mod vendor` step is pretty quick if all the modules are already downloaded into the cache.

stmcginnis replied:
Does adding `rerun-if-changed` help here? Like rust's `Cargo.lock` file, the `go.mod` and `go.sum` files "lock" dependencies to a specific version. So there shouldn't be any change in what `go mod vendor` would download unless there are changes to those files.

At least I'm pretty sure. I've never mixed `go.mod` module management with vendoring in a project before, but this is the way I understand it to work.

jpmcb replied:
> it might be simpler to just always recreating the bundle file ... the go mod vendor step is pretty quick if all the modules are already downloaded into the cache.

I think recreating the dependency bundle is the way to go for now and should be almost instant given the modcahce, but I'll also do some experimenting to see if it slows down compared to my testing yesterday.

bcressey replied:
> Does adding `rerun-if-changed` help here? Like rust's Cargo.lock file, the go.mod and go.sum files "lock" dependencies to a specific version.

It sort of helps, in the sense that any change to the file would trigger a rebuild - but if we're in `buildsys` at all then that means a (re)build has been triggered, and we don't know exactly why, just that `cargo` decided it was necessary. Since we're doing our own caching here, we'd need to detect whether the original input file changed - something else might've changed instead - and invalidate the bundled archive if it did.

That could be done by also saving a hash file of the original file somewhere, but then we have to manage that file too. Or we could prepend the hash of the original file to the bundled archive filename, so we could always check it, but then we'd want to strip off that prefix before passing into the package build environment since it wouldn't be predictable.

Not insurmountable problems but if the added benefit of the secondary cache is minor, it might not be worth the complexity.

jpmcb replied:
Ended up going with rebuilding the vendor directory and re-archiving. It didn't have an impact on performance since the local modcache should already be populated with the dependencies. It also seemed overly complicated to track the hashes of the dependencies and force a rebuild if any of those change.

But, like Sean said, I'd be very very surprised if those dependencies _did_ change out of band from the upstream releasing a new version of their software: the `go.mod` and `go.sum` act as locks on what dependencies are needed and shouldn't be different build to build.

stmcginnis commented:
I've found some issues with using `go mod vendor` for pulling in dependencies for projects that do not using vendoring for dependency management. Not sure if there's a better way for us to handle it than current plans, but thought I should add my findings here so others are aware and we can possibly think about a better approach.

## Vendoring approach

I found these issues as I'm trying to build the `ecr-credential-provider` binary out of https://github.com/kubernetes/cloud-provider-aws. If I try to build it isolated (airgapped, offline, etc), I first run `go mod vendor` to create and populate a vendor directory. But then when I try to build the project, it appears the vendor directory is too deep or something, so I get this error:

```shell
$ make ecr-credential-provider
GO111MODULE=on CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GOPROXY=direct go build \
	-trimpath \
	-ldflags="-w -s -X k8s.io/component-base/version.gitVersion=v1.22.3-dirty" \
	-o=ecr-credential-provider \
	cmd/ecr-credential-provider/*.go
make: /bin/bash: Argument list too long
make: *** [Makefile:38: ecr-credential-provider] Error 127
```

## `go mod download` Approach

This led me to explore a more module-centric solution to capturing all deps to be able to build offline.

I first tried to capture dependencies into a separate directory so I could use that to copy into the offline build.

```shell
GOCACHE="$(mktemp -d)"
GOPATH="$(mktemp -d)"

GOPATH="${GOPATH}" go mod download
cp -rp "${GOPATH}"/pkg/mod/cache/download/* "${GOCACHE}/"

docker run --rm --network none \
    -v "${GOCACHE}":/gocache \
    -v ${PWD}:/build -w /build \
    -e GOCACHE=/cache \
    -e GOPROXY=file:///gocache golang:1.18 \
    make ecr-credential-provider
```

This had a couple issues. For some reason, more often than not the `go mod download` would exit with failure being unable to find one of the dependencies:

```shell
go: github.com/matttproud/golang_protobuf_extensions@v1.0.2-0.20181231171920-c182affec369 requires
	golang.org/x/sync@v0.0.0-20181221193216-37e7f081c4d4: invalid version: unknown revision 37e7f081c4d4
```

If I reran the command a few times, eventually it would succeed and get everything downloaded.

But the second issue then was that `go mod download` doesn't go deep. The downloaded dependencies wouldn't actually have everything needed to build the project, so running in an offline environment would error out stating there were things missing:

```
...
cmd/ecr-credential-provider/main.go:32:2: github.com/aws/aws-sdk-go@v1.43.32: reading file:///gocache/github.com/aws/aws-sdk-go/@v/v1.43.32.zip: no such file or directory
cmd/ecr-credential-provider/main.go:34:2: k8s.io/apimachinery@v0.22.8: reading file:///gocache/k8s.io/apimachinery/@v/v0.22.8.zip: no such file or directory
cmd/ecr-credential-provider/plugin.go:28:2: k8s.io/apimachinery@v0.22.8: reading file:///gocache/k8s.io/apimachinery/@v/v0.22.8.zip: no such file or directory
...
```

## Double Build Approach

I was able to get everything downloaded by first building the project while still connected:

```
GOPATH="${GOPATH}" GOPROXY="file:///${GOCACHE}" make ecr-credential-provider
```

then building again isolated:

```
rm ecr-credential-provider

docker run --rm --network none \
    -v "${GOCACHE}":/gocache \
    -v ${PWD}:/build -w /build \
    -e GOCACHE=/cache \
    -e GOPROXY=file:///gocache golang:1.18 \
    make ecr-credential-provider
```

It is a small project that doesn't take long to build, but that seems a little redundant. If we did end up having this codified in the way buildsys handles go projects, this definitely would not be a good thing to do for all go dependencies.

## "Thorough" Approach

What ended up working the best was to "force" `go mod download` into actually following all dependencies. This is what I ended up with that would consistently work:

```
GOCACHE="$(mktemp -d)"
GOPATH="$(mktemp -d)"

go mod graph | \
    awk '{print $2}' | \
    sort -u | \
    GOPATH="${GOPATH}" xargs go mod download
cp -rp "${GOPATH}"/pkg/mod/cache/download/* "${GOCACHE}/"

docker run --rm --network none \
    -v "${GOCACHE}":/gocache \
    -v ${PWD}:/build -w /build \
    -e GOCACHE=/cache \
    -e GOPROXY=file:///gocache golang:1.18 \
    make ecr-credential-provider
```

The downside is the `go mod graph` call does take some time since it is actually thorough. I think we could mitigate that by having a local (non-isolated) GOCACHE that buildsys could use while downloading the dependencies. Then we shouldn't need to download the same packages repeatedly.

jpmcb commented:
Added another commit to address latest review comments:

Now, when `buildsys` vendors the go dependencies, it places them into a separate archive. The pristine upstream archive can then be used in conjunction with the dependency archive to create the package.

Adds sane defaults for several of the bundling options.

---

I will plan to squash this into a co-authored commit upon approval, but this was a bit more of a change that seemed to warrant it's own intermittent commit

jpmcb commented:
@stmcginnis - I think something is wrong with the cloud-provider-aws `v1.22.3` tag and they have an out of date `go.mod` file.

---

I am able to pull down the tar with
```
❯ wget https://github.com/kubernetes/cloud-provider-aws/archive/refs/tags/v1.22.3.tar.gz
```

and unpack it with:
```
❯ tar -zxvf v1.22.3.tar.gz
```

Then, going into that directory, I can vendor the code. But when I attempt to manually run a `go build` I get the following:
```
cd cloud-provider-aws-1.22.3
go mod vendor
GO111MODULE=on CGO_ENABLED=0 GOOS=linux GOARCH=arm64 GOPROXY=https://proxy.golang.org,direct go build -x \
        -trimpath \
        -ldflags="-w -s -X k8s.io/component-base/version.gitVersion=" \
        -o=ecr-credential-provider \
        cmd/ecr-credential-provider/*.go
WORK=/tmp/go-build4216558415
go: updates to go.mod needed, disabled by -mod=vendor
        (Go version in go.mod is at least 1.14 and vendor directory exists.)
        to update it:
        go mod tidy
```

which is super weird. It's asking me to run a `go mod tidy`? I'd expect that to "just work" and there not need to be an update to the module dependencies. Switching over to the actual repository, I checked out their `v1.22.3` tag and see the same problem with the command above.

So, running `go mod tidy` on their git repo _does_ make it dirty and show the dependencies are out of date:
```
diff --git a/go.mod b/go.mod
index 27a8a4a..b19147b 100644
--- a/go.mod
+++ b/go.mod
@@ -6,6 +6,7 @@ require (
        github.com/aws/aws-sdk-go v1.43.32
        github.com/golang/mock v1.4.4
        github.com/google/go-cmp v0.5.5
+       github.com/spf13/pflag v1.0.5
        github.com/stretchr/testify v1.7.0
        gopkg.in/gcfg.v1 v1.2.0
        k8s.io/api v0.22.8
@@ -66,7 +67,6 @@ require (
        github.com/prometheus/common v0.26.0 // indirect
        github.com/prometheus/procfs v0.6.0 // indirect
        github.com/spf13/cobra v1.1.3 // indirect
-       github.com/spf13/pflag v1.0.5 // indirect
        github.com/stretchr/objx v0.2.0 // indirect
        go.etcd.io/etcd/api/v3 v3.5.0 // indirect
        go.etcd.io/etcd/client/pkg/v3 v3.5.0 // indirect
@@ -106,7 +106,6 @@ require (
        gopkg.in/yaml.v2 v2.4.0 // indirect
        gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b // indirect
        k8s.io/apiserver v0.22.8 // indirect
-       k8s.io/controller-manager v0.22.8 // indirect
        k8s.io/gengo v0.0.0-20201214224949-b6c5ce23f027 // indirect
        k8s.io/kube-openapi v0.0.0-20211109043538-20434351676c // indirect
        sigs.k8s.io/apiserver-network-proxy/konnectivity-client v0.0.30 // indirect
```

The `pflag` one is especially interesting since it _does_ seems they use it directly in their controller:

```
❯ rg pflag
...

pkg/controllers/options/tagging_controller.go
18:     "github.com/spf13/pflag"
29:func (o *TaggingControllerOptions) AddFlags(fs *pflag.FlagSet) {
```

But checking out their `master` branch, cleaning everything up, and running through the sequence again works fine without any problem in their `go.mod`:

```
❯ git checkout master

❯ rm ecr-credential-provider && rm -rf vendor

❯ git reset --hard origin/master
HEAD is now at c72f89b Merge pull request #471 from wangamel/master

❯ GO111MODULE=on CGO_ENABLED=0 GOOS=linux GOARCH=arm64 GOPROXY=https://proxy.golang.org,direct go build -x \
        -trimpath \
        -ldflags="-w -s -X k8s.io/component-base/version.gitVersion=" \
        -o=ecr-credential-provider \
        cmd/ecr-credential-provider/*.go

❯ ll ecr-credential-provider
-rwxr-xr-x. 1 fedora fedora 14M Aug 26 22:53 ecr-credential-provider
```

---

My presumption is their `v1.22.3` tag had an out of date go.mod which causes `go mod vendor` to get the wrong dependencies which then prevents a build.

I also wonder if `make` was eating all of that error output as an argument list?

Can we ask them to cut a new, up to date tag / tar with the right go.mod?

stmcginnis commented:
> My presumption is their v1.22.3 tag had an out of date go.mod which causes go mod vendor to get the wrong dependencies which then prevents a build.

Go would update the go.mod file, so I had assumed that would resolve any issues before vendoring. But I think you're right there. I found there is a newer v1.24.1. tag and the dependency issues have since been cleaned up (with one minor change still stale).

That seems to be enough that it fixed whatever was happening though:

```
$ git branch
* (HEAD detached at v1.24.1)
  master
$ go mod tidy
$ git status
HEAD detached at v1.24.1
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   go.mod

no changes added to commit (use "git add" and/or "git commit -a")
$ PAGER= git diff
diff --git a/go.mod b/go.mod
index 76e3379..8c6d8ca 100644
--- a/go.mod
+++ b/go.mod
@@ -8,6 +8,7 @@ require (
 	github.com/google/go-cmp v0.5.8
 	github.com/spf13/pflag v1.0.5
 	github.com/stretchr/testify v1.7.2
+	golang.org/x/time v0.0.0-20220210224613-90d013bbcef8
 	gopkg.in/gcfg.v1 v1.2.3
 	k8s.io/api v0.24.2
 	k8s.io/apimachinery v0.24.2
@@ -94,7 +95,6 @@ require (
 	golang.org/x/sys v0.0.0-20220209214540-3681064d5158 // indirect
 	golang.org/x/term v0.0.0-20210927222741-03fcf44c2211 // indirect
 	golang.org/x/text v0.3.7 // indirect
-	golang.org/x/time v0.0.0-20220210224613-90d013bbcef8 // indirect
 	golang.org/x/tools v0.1.10-0.20220218145154-897bd77cd717 // indirect
 	golang.org/x/xerrors v0.0.0-20200804184101-5ec99f83aff1 // indirect
 	google.golang.org/appengine v1.6.7 // indirect
$ go mod vendor 
$ docker run --rm --network none -v ${PWD}:/build -w /build golang:1.18 make ecr-credential-provider
GO111MODULE=on CGO_ENABLED=0 GOOS=linux GOARCH=arm64 GOPROXY=https://proxy.golang.org,direct go build \
	-trimpath \
	-ldflags="-w -s -X k8s.io/component-base/version.gitVersion=v1.24.1-dirty" \
	-o=ecr-credential-provider \
	cmd/ecr-credential-provider/*.go
$ ls ecr-credential-provider
ecr-credential-provider
```

Glad this is at least captured here as I think we may run into other weirdness like this as we add other go packages. But looks like I can at least proceed for now.

And this change looks good to me. Again, just wanted to capture some weirdness so there's some record of it.

jpmcb commented:
> Glad this is at least captured here as I think we may run into other weirdness like this as we add other go packages. But looks like I can at least proceed for now.

@stmcginnis Awesome!!! 👏🏼 glad that works!! And yes, glad it's captured here

stmcginnis reviewed (approved):
Tested this out with the work I'm doing. I'm still working through other issues, but I've gotten past the part covered by this and everything appears to be working well.

bcressey commented on packages/hotdog/hotdog.spec:
```diff
@@ -10,67 +10,44 @@
-cp GOPATH/src/golang.org/x/sys/LICENSE LICENSE.golang-sys
+# Expand the upstream source code tarball
+# -n: use the named directory "hotdog-{long-sha}"
+%setup -n %{gorepo}-%{gitrev}
```
This level of detail is helpful in the ecs-agent spec, which has so much going on that it'd be inscrutable without the comments.

Here I think it'd be simpler as a two-liner:
```suggestion
%setup -n %{gorepo}-%{gitrev}
%setup -T -D -n %{gorepo}-%{gitrev} -b 1
```

Otherwise it's a bit like trying to teach the language through comments.

jpmcb replied:
> Otherwise it's a bit like trying to teach the language through comments.

Yeah, super fair. I wasn't sure what level of detail to go into with what was happening here (and was looking at a bunch of the other specs to see what was happening) but yes, a simple 2 liner makes sense!

bcressey commented on tools/buildsys/src/manifest.rs:
```diff
@@ -33,6 +33,46 @@ url = "https://bar"
+
+`bundle-path` is an optional argument that provides the filepath
+within the archive that contains the module.
+By default, the first top level directory in the archieve is used.
```
```suggestion
By default, the first top level directory in the archive is used.
```

bcressey commented on tools/buildsys/src/manifest.rs:
```diff
@@ -33,6 +33,46 @@ url = "https://bar"
+this `bundle-path` value should be `a/b/c`.
+Or, given a archieve with a single directory that contains a go project
+that has go.mod and go.sum files, this option may be omitted since the single
+top level directory will be used.
```
nit: various formatting suggestions
```suggestion
So, for example, given a Go project that has the necessary `go.mod` and `go.sum`
files within the archive located at the filepath `a/b/c`,
this `bundle-path` value should be "a/b/c".
Or, given an archive with a single directory that contains a Go project
that has `go.mod` and `go.sum` files, this option may be omitted since the single
top-level directory will be used.
```

bcressey commented on tools/buildsys/src/manifest.rs:
```diff
@@ -33,6 +33,46 @@ url = "https://bar"
+but pre-pended with "bundled-".
+This path may be referenced within a spec, when creating a package, etc.
+For example, `my-package-bundled.tar.gz` may be entered as an output path
+indicating the package is "bundled" and has been archived.
```
I don't quite follow the second part of this sentence.

It might be better to reorder the sentences somewhat and talk about usage in the spec file at the end.

bcressey commented on tools/buildsys/src/manifest.rs:
```diff
@@ -33,6 +33,46 @@ url = "https://bar"
+url = "https://foo"
+sha512 = "abcdef"
+bundle-modules = [ "go" ]
+bundle-path = "path/to/module/"
```
nit: unless the trailing slash is significant (which it shouldn't be) I'd omit it so that people don't feel like they always have to include it.

Maybe the field should be `bundle-control-path` or something, to better distinguish it from `bundle-output-path`?

jpmcb replied:
Hmmmm good point. Naming things is hard. What about `bundle-root-path` to indicate it's the root of the bundle where the module's dependency tracking occurs?

bcressey commented on tools/docker-go:
```diff
@@ -56,7 +56,7 @@ parse_args "${@}"
 # Pass through relevant Go variables, from the config or environment.
 go_env=( )
-for i in GOPROXY GONOPROXY GOPRIVATE ; do
+for i in GOPROXY GONOPROXY GOPRIVATE GOSUMDB ; do
```
Nice, this is much simpler.

bcressey commented on tools/buildsys/src/main.rs:
```diff
@@ -143,6 +150,20 @@ fn build_package() -> Result<()> {
+                continue;
+            }
+
+            for b in f.bundle_modules.as_ref().unwrap() {
+                println!("cargo:rerun-if-changed={:?}", f.bundle_output_path);
```
This doesn't do the right thing if the output path is left unspecified, right? Since we haven't calculated the default value yet. But I think it's redundant with the logic in `gomod.rs`.

jpmcb replied:
Ah yes, correct; it is redundant - this was a leftover from the first commit.

Removing it.

bcressey commented on tools/buildsys/src/gomod/error.rs:
```diff
@@ -0,0 +1,33 @@
+    CommandStart { source: std::io::Error },
+
+    #[snafu(display("Failed to execute command: 'docker {:?}'", args))]
+    DockerExecution { args: String },
```
`String` implements `Display`, so you don't have to use the `Debug` impl:
```suggestion
    #[snafu(display("Failed to execute command: 'docker {}'", args))]
    DockerExecution { args: String },
```

bcressey commented on tools/buildsys/src/gomod/error.rs:
```diff
@@ -0,0 +1,33 @@
+    InputFile,
+
+    #[snafu(display("Input file {:?} must be a file", path))]
+    InputFileBad { path: PathBuf },
```
```suggestion
    #[snafu(display("Input file {} must be a file", path.display()))]
    InputFileBad { path: PathBuf },
```

bcressey commented on tools/buildsys/src/gomod.rs:
```diff
@@ -0,0 +1,151 @@
+Packages using the Go programming language may have upstream tar archives that
+include only the source code of the project, but not the source code of any
+dependencies. The Go programming language promotes the use of "modules" for
+dependencies and projects adopting modules will provide go.mod and go.sum
```
nit:
```suggestion
dependencies and projects adopting modules will provide `go.mod` and `go.sum`
```

(same nit below)

bcressey commented on tools/buildsys/src/gomod.rs:
```diff
@@ -0,0 +1,151 @@
+The location where dependencies are retrieved from are controlled by the
+standard environment variables employed by the Go tool: GOPROXY, GOSUMDB, and
+GOPRIVATE and are automatically retrieved from the host environment
+when docker-go is envoked.
```
```suggestion
The location where dependencies are retrieved from are controlled by the
standard environment variables employed by the Go tool: `GOPROXY`, `GOSUMDB`, and
`GOPRIVATE`. These variables are automatically retrieved from the host environment
when docker-go is invoked.
```

bcressey commented on tools/buildsys/src/gomod.rs:
```diff
@@ -0,0 +1,151 @@
+        gomod: &manifest::ExternalFile,
+    ) -> Result<()> {
+        let url_file_name = extract_file_name(&gomod.url)?;
+        let path_arg = &gomod.path.as_ref().unwrap_or(&url_file_name);
```
nit:
```suggestion
        let local_file_name = &gomod.path.as_ref().unwrap_or(&url_file_name);
```

bcressey commented on tools/buildsys/src/gomod.rs:
```diff
@@ -0,0 +1,151 @@
+    pub(crate) fn vendor(
+        root_dir: &Path,
+        package_dir: &Path,
+        gomod: &manifest::ExternalFile,
```
Maybe this for clarity?
```suggestion
        external_file: &manifest::ExternalFile,
```

bcressey commented on tools/buildsys/src/gomod.rs:
```diff
@@ -0,0 +1,151 @@
+
+        // If a module directory was not provided, set as an empty path.
+        // By default, without a provided module directory, tar will be passed
+        // the first directory found in the archieve as the top level Go module
```
```suggestion
        // the first directory found in the archive as the top level Go module
```

bcressey commented on tools/buildsys/src/gomod.rs:
```diff
@@ -0,0 +1,151 @@
+        // By default, without a provided module directory, tar will be passed
+        // the first directory found in the archieve as the top level Go module
+        let default_empty_path = PathBuf::from("");
+        let mod_dir = gomod.bundle_path.as_ref().unwrap_or(&default_empty_path);
```
```suggestion
        let mod_dir = gomod.bundle_path.as_ref().unwrap_or_else(|| "".to_string());
```

jpmcb replied:
I had tried this originally but got hung up on the typing:

```
error[E0308]: mismatched types
  --> buildsys/src/gomod.rs:48:76
   |
48 |         let mod_dir = external_file.bundle_path.as_ref().unwrap_or_else(|| "".to_string()); // &default_empty_path);
   |                                                                            ^^^^^^^^^^^^^^ expected `&PathBuf`, found struct `std::string::String`
```

My understanding is that since `external_file_path` is a `PathBuf` and we are attempting to unwrap it, we would also need to return a pathbuf if it fails, not just an empty string? That's what lead me down the path of declaring an empty pathbuf and using it's borrowed value as the default if none was provided.

bcressey commented on tools/buildsys/src/gomod.rs:
```diff
@@ -0,0 +1,151 @@
+                go list -mod=readonly ./... >/dev/null && go mod vendor &&
+                popd &&
+                tar czf {output} $(if [ -z {moddir} ]; then echo $toplevel; else echo {moddir}; fi)/vendor && 
+                rm -rf $(if [ -z {moddir} ]; then echo $toplevel; else echo {moddir}; fi)",
```
This might be more readable as a constant toward the top of the program, using a multi-line raw string literal.

It might also be worth defining a function (like `go_fmt` or `go_fetch` in `Makefile.toml` so it's more clear how the parameters we pass into the format string line up with the parts of the execution that vary.

bcressey commented on tools/buildsys/src/gomod.rs:
```diff
@@ -0,0 +1,151 @@
+                // Finally, it cleans up by removing the untar'd source code.
+                // The upstream archive remains and both tar files can be used during packaging.
+                "toplevel=$(tar tf {input} | head -1)
+                tar xf {input} &&
+                pushd $(if [ -z {moddir} ]; then echo $toplevel; else echo {moddir}; fi) &&
```
I'd recommend not doing the Dockerfile-ish technique of writing `&&` everywhere if it's not necessary.

This would be more readable as:
```
if [ -n "${moddir}" ] ; then
  targetdir="${moddir}"
else
  targetdir="${toplevel}"
fi

pushd "${targetdir}"
# do the vendoring
popd

if [ ! -d "${targetdir}/vendor ] ;
  echo "no vendor directory found" >&2
  exit 1
fi

tar czf ... "${targetdir}/vendor"

# clean up - why?
rm -rf "${targetdir}"
```

bcressey reviewed (commented):
Nice work! Lots of nits but I don't see anything major missing.

jpmcb commented:
Added _another_ commit which reworks how the bundling script is used by `docker-go`:

Now, instead of attempting to feed a long script delimited by `&&` or `\`, the script lives as a `const` string that gets added as a temporary file to the package directory. This file is then accessible by the container and easily run with `bash -c ./docker-go-script.sh`. Further, the bundling script now expects several environment variables which are loaded through a new `--additional-docker-args` option in the `docker-go` tool; this enables us to use the `-e` arguments for `docker` to load up environment variables into the container.

Further documentation in comments in code.

stmcginnis commented on tools/docker-go:
```diff
@@ -18,6 +18,9 @@ Required:
     --command                   The command to run in the SDK container
+
+Optional:
+	--additional-docker-args    Additional free-form arguments to provide to "docker run" (i.e. mounts, env-vars, etc.)
```
Indentation is a little off here. Minor thing if you need to do another update.

jpmcb replied:
Hmmm weird, nvim inserted a tab here instead of 4 spaces. But yes, good catch! Will update with other feedback

bcressey commented on tools/buildsys/src/manifest.rs:
```diff
@@ -33,6 +33,49 @@ url = "https://bar"
+and are used to untar an upstream external file archive,
+vendor any dependent code,
+and produce an additional archive with those dependencies.
+Only `bundle-modules` is required when bundling an archive's dependences.
```
nit: the line-wrapping seems pretty short or inconsistent here, not that it was very consistent in this file before. I'd say either 80 or 100 are acceptable values. It's more noticeable down by the `bundle-output-path` comments.

bcressey commented on tools/buildsys/src/gomod.rs:
```diff
@@ -0,0 +1,213 @@
+tar xf "${LOCAL_FILE_NAME}"
+
+pushd "${targetdir}"
+	go list -mod=readonly ./... >/dev/null && go mod vendor
```
nit: a tab snuck in here
```suggestion
    go list -mod=readonly ./... >/dev/null && go mod vendor
```

bcressey commented on tools/buildsys/src/gomod.rs:
```diff
@@ -0,0 +1,213 @@
+                fs::remove_file(script_path).unwrap();
+                Err(e)
+            }
+        }
```
Hmm, a more idiomatic way to do this might be:
```suggestion
        let res = docker_go(root_dir, &args);
        fs::remove_file(script_path)?;
        res
```

bcressey commented on tools/buildsys/src/gomod.rs:
```diff
@@ -0,0 +1,213 @@
+            sdk_image: sdk,
+            go_mod_cache: &root_dir.join(".gomodcache"),
+            command: format!("./{}", GO_MOD_DOCKER_SCRIPT_NAME),
+            extra_docker_args: vec![
+                // Sets container environment variables needed for the docker-go-script
+                format!("-e LOCAL_FILE_NAME={}", local_file_name.to_string_lossy()),
+                format!("-e MOD_DIR={}", mod_dir.to_string_lossy()),
+                format!("-e OUTPUT={}", output_path_arg.to_string_lossy()),
+            ],
```
I don't completely love the new "extra docker args" functionality since then it's harder to reason about what the tool will or won't do.

Rather than use variables in the script, it might be worth using placeholders like `__LOCAL_FILE_NAME__` and then just replacing them with these values before writing out the file.

jpmcb replied:
Good call - this was likely an implementation artifact from exploring how a script could be used with the `docker-go` tool. I think using a template string is less complex in the end and if someone wanted to look at the script afterwards (or while in process) they could do see those values populated

bcressey reviewed (commented):
Looks good. In terms of the commits:
* I'd "unsquash" the first one and have two separate commits for the `hotdog` and `oci-add-hooks` changes.
* There's at least one typo in the second commit message
* You can squash all the `buildsys` commits together as the first patch in the series (probably you were planning to do this).

jpmcb committed:
buildsys: extend external-files to vendor go modules

Using the `bundle-*` keys on
`packag.metadata.build-package.external-files`, buildsys can vendor Go
dependencies into a new archive (leaving the prestine upstream archive
intact). Both archives may then be used to build the package in an
isolated environment.

toosl/docker-go now also retrieves the `GOSUMDB` environment variable
from the host.

Co-authored-by: John McBride <jpmmcb@amazon.com>
Co-authored-by: Samuel Karp <skarp@amazon.com>
Signed-off-by: John McBride <jpmmcb@amazon.com>

jpmcb committed:
hotdog: use new buildsys go module support

Co-authored-by: John McBride <jpmmcb@amazon.com>
Co-authored-by: Samuel Karp <skarp@amazon.com>
Signed-off-by: John McBride <jpmmcb@amazon.com>

jpmcb committed:
oci-add-hooks: use new buildsys go module support

Co-authored-by: John McBride <jpmmcb@amazon.com>
Co-authored-by: Samuel Karp <skarp@amazon.com>
Signed-off-by: John McBride <jpmmcb@amazon.com>

jpmcb commented:
Force pushed to flesh out the commit structure per @bcressey's recommendations and address a few nits!

Removed changes to `tools/docker-go` which added the `--additional-docker-args` and instead, the go module script uses a few placeholder variables that are populated before the temporary file is put in the package directory.

bcressey reviewed (approved):
Yay!

stmcginnis reviewed (approved):
Tested with https://github.com/bottlerocket-os/bottlerocket/pull/2377 and it looks like everything works as expected.
